_posts/2022-01-31-how-we-ship-ios-libraries.md

```diff
@@ -0,0 +1,333 @@
+
+```
+; ls -l target/debug/libshipping_rust_ffi.a
+-rw-r--r-- 2 jer staff 16061952 Jan 28 13:09 target/debug/libshipping_rust_ffi.a
+```
+
+Let's add some code to `shipping-rust-ffi/src/lib.rs` next.
+Nothing fancy, a simple function taking some arguments and returning the sum:
+
+```rust
+use std::os::raw::c_int;
+
+#[no_mangle]
+pub extern "C" fn shipping_rust_addition(a: c_int, b: c_int) -> c_int {
+    a + b
+}
+```
+
+The `no_mangle` ensures the name lands in the compiled library as is
+and the `extern "C"` makes sure it uses the right ABI.
+
+We now got a Rust library, exporting a C-ABI compatible interface.
+We can now consume this in our iOS application.
+
+## The Xcode parts
+
+Before we can use the code we need a bit more setup.
+
+We start by linking against the `libshipping_rust_ffi.a` library.
+In your Xcode project open your target configuration[^1],
+go to "Build Phases", then look for "Link Binary with Libraries".
+Add a new one, in the popup select "Add files" on the bottom left
+and look for the `target/debug/libshipping_rust_ffi.a` file.
+Yes, that's actually for the wrong target. This is just for the name, we'll fix up the path next.
+Go to "Build Settings" and search for "Library Search Paths".
+It probably has the path to file in there right now for both `Debug` and `Release` builds.
+Remove that one for `Debug`, then add a new row by clicking the small `+` symbol.
+Select any matcher, e.g. `Any Driverkit`. It doesn't matter what value you give it.
+We will manually modify that entry next.
+Do the same for the `Release` configuration.
+
+Once that's done, save your project and go back to your project directory.
+We will modify the project configuration to have Xcode look for the library based on the target it is building for[^2].
+Open up `ShippingRust.xcodeproj/project.pbxproj` in a text editor,
+then search for the first line with `"LIBRARY_SEARCH_PATHS[sdk=driverkit*]"`.
+It should be in a section saying `/* Debug */`.
+Remove that line and add 3 new ones:
+
+```
+"LIBRARY_SEARCH_PATHS[sdk=iphoneos*][arch=arm64]" = "$(PROJECT_DIR)/target/aarch64-apple-ios/debug";
+"LIBRARY_SEARCH_PATHS[sdk=iphonesimulator*][arch=arm64]" = "$(PROJECT_DIR)/target/aarch64-apple-ios-sim/debug";
+"LIBRARY_SEARCH_PATHS[sdk=iphonesimulator*][arch=x86_64]" = "$(PROJECT_DIR)/target/x86_64-apple-ios/debug";
+```
+
+Look for the next line with `"LIBRARY_SEARCH_PATHS[sdk=driverkit*]"`, now in a `/* Release */` section and replace it with:
+
+```
+"LIBRARY_SEARCH_PATHS[sdk=iphoneos*][arch=arm64]" = "$(PROJECT_DIR)/target/aarch64-apple-ios/release";
+"LIBRARY_SEARCH_PATHS[sdk=iphonesimulator*][arch=arm64]" = "$(PROJECT_DIR)/target/aarch64-apple-ios-sim/release";
+"LIBRARY_SEARCH_PATHS[sdk=iphonesimulator*][arch=x86_64]" = "$(PROJECT_DIR)/target/x86_64-apple-ios/release";
+```
+
+Save the file and return back to Xcode.
+If you didn't make any typos Xcode should still be able to open your project.
```

Comment:
"should still be able to open your project" suggests you should close the project and open it up again in Xcode. Is that something you should do here? If you don't do that, are there other indicators you succeeded/failed in the previous steps?

Reply:
Xcode shouting at you "Can't open this project" (while it's still open but can't open the settings anymore).

I'll make that more clear in the text.